applications/GeoMechanicsApplication/custom_elements/U_Pw_small_strain_element.cpp

```diff
@@ -1254,80 +1252,72 @@ void UPwSmallStrainElement<TDim, TNumNodes>::CalculateSoilGamma(ElementVariables
 {
     KRATOS_TRY
 
-    rVariables.Density = GeoTransportEquationUtilities::CalculateSoilDensity(
-        rVariables.DegreeOfSaturation, this->GetProperties());
-
-    noalias(rVariables.SoilGamma) = rVariables.Density * rVariables.BodyAcceleration;
+    noalias(rVariables.SoilGamma) = GeoTransportEquationUtilities::CalculateSoilDensity(
+                                        rVariables.DegreeOfSaturation, this->GetProperties()) *
+                                    rVariables.BodyAcceleration;
 
     KRATOS_CATCH("")
 }
 
 template <unsigned int TDim, unsigned int TNumNodes>
 void UPwSmallStrainElement<TDim, TNumNodes>::CalculateAndAddCouplingTerms(VectorType& rRightHandSideVector,
-                                                                          ElementVariables& rVariables)
+                                                                          const ElementVariables& rVariables)
 {
     KRATOS_TRY
 
-    rVariables.UPMatrix = (-1.0) * GeoTransportEquationUtilities::CalculateCouplingMatrix(
-                                       rVariables.B, this->GetStressStatePolicy().GetVoigtVector(),
-                                       rVariables.Np, rVariables.BiotCoefficient,
-                                       rVariables.BishopCoefficient, rVariables.IntegrationCoefficient);
-    rVariables.UVector = prod(rVariables.UPMatrix, rVariables.PressureVector);
-    GeoElementUtilities::AssembleUBlockVector(rRightHandSideVector, rVariables.UVector);
+    const BoundedMatrix<double, TNumNodes * TDim, TNumNodes> coupling_matrix =
+        (-1.0) * GeoTransportEquationUtilities::CalculateCouplingMatrix(
+                     rVariables.B, this->GetStressStatePolicy().GetVoigtVector(), rVariables.Np,
+                     rVariables.BiotCoefficient, rVariables.BishopCoefficient, rVariables.IntegrationCoefficient);
+
+    const array_1d<double, TNumNodes * TDim> coupling_terms = prod(coupling_matrix, rVariables.PressureVector);
+    GeoElementUtilities::AssembleUBlockVector(rRightHandSideVector, coupling_terms);
 
     if (!rVariables.IgnoreUndrained) {
         const double SaturationCoefficient = rVariables.DegreeOfSaturation / rVariables.BishopCoefficient;
-        rVariables.PVector = PORE_PRESSURE_SIGN_FACTOR * SaturationCoefficient *
-                             prod(trans(rVariables.UPMatrix), rVariables.VelocityVector);
-        GeoElementUtilities::AssemblePBlockVector(rRightHandSideVector, rVariables.PVector);
+        const array_1d<double, TNumNodes> undrained_terms =
```

Comment:
Same here, the coupling flow is computed. Undrained is a confusing description.

Reply:
Now it is `transposed_coupling_matrix`